### DISCUSSION TITLE
Unhandled Runtime Error MongoDB

### DISCUSSION THREAD

**Discussion by Mona904** [neuefische/questions#607] [Web] [answered]
Hello all, 

I have tried to connect my MongoDB Database with VSCode but to no avail. The error message I get: 
<img width="712" alt="Screenshot 2023-05-17 at 16 07 40" src="https://github.com/neuefische/questions/assets/126586287/e4d9319e-4bed-4faa-ac4e-b49054791c0f">

I have almost the exact same code as one of the other students with whom I have worked together on this. 
Please advise. Here is my code:

`import mongoose, { model, models, Schema } from "mongoose";

 const { MONGODB_URL } = process.env;

 const PlantSchema = new Schema ({
  title: String,
  location: String,
 });

 const Plant = models.Plant || model("Plant", PlantSchema);

 async function connectDatabase() {
  await mongoose.connect(MONGODB_URL);
}

async function createPlant(plant) {
  await connectDatabase();
  const newPlant = await Plant.create(plant);
  return newPlant;
}

async function getPlant(id) {
  await connectDatabase();
  const plant = await Plant.findOne({ _id: id });
  return plant;
}

async function getAllPlants() {
  await connectDatabase();
  const elements = await Plant.find({});
  return elements;
}

async function deletePlant(id) {
  await connectDatabase();
  const plant = await Plant.findOne({ _id: id });
  await Plant.deleteOne({ _id: id });
  return plant;
}

async function updatePlant(id, plant) {
  await connectDatabase();
  await Plant.updateOne({ _id: id }, plant);
  const updatedPlant = await Plant.findOne({ _id: id });
  return updatedPlant;
}

export default Plant;

export {createPlant, updatePlant, deletePlant, getAllPlants, getPlant};`

**Comment by mbosselmann** [comment]
Hi Mona,
can you give us a link to your current code branch? Otherwise we cannot check your problem. :)

**Reply by Mona904** [reply]
Yes just a second!

**Reply by Mona904** [reply]
[https://github.com/Mona904/capstone-project/pull/5](PR)

**Reply by mbosselmann** [reply, marked as answer]
Hi Mona, 
I checked your code and I think it is not the right solution to work with a database at this point. I think you wanted to share your plant state with your index page and your favorites page and I would refactor your code to handle this. The keyword is lifting up state. 
On my way through your repo I found some more errors that I tried to solve and that I did not solve yet. So here is my path for you to try out. If this does not work out let us discuss on Friday. :)

1. ´db.js` and `pages/api/Plants/index.js`
- I deleted `pages/api/Plants/index.js`
- instead of putting the database data here I put your plant data in there:
```js
export const plantData = [
  {
    id: 1,
    name: "Plant 10",
    image:
      "https://www.picturethisai.com/wiki-image/1080/154128268713263129.jpeg",
  },
  {
    id: 2,
    name: "Plant 20",
    image:
      "https://www.picturethisai.com/image-handle/website_cmsname/image/1080/154244786881036288.jpeg?x-oss-process=image/format,webp/resize,s_422&v=1.0",
  },
  {
    id: 3,
    name: "Plant 30",
    image:
      "https://cdn.shopify.com/s/files/1/0557/0657/products/Quercus-robur.jpg",
  },
];
```
2. `_app.js`
- in your _app.js I imported the data and saved them as initial data in a state:
- I also moved your toggleFavorite function to your _app.js and refactored it as it was not working correctly
- I handed over `plants` and `toggleFavorite` as props

```jsx
import GlobalStyle from "../styles";
import { plantData } from "@/db";
import { useState } from "react";

export default function App({ Component, pageProps }) {
  const [plants, setPlants] = useState(plantData);

  const toggleFavorite = (plantId) => {
    setPlants(
      plants.map((plant) => {
        if (plant.id === plantId) {
          return {
            ...plant,
            isFavorite: plant.isFavorite ? !plant.isFavorite : true,
          };
        } else {
          return plant;
        }
      })
    );
  };

  return (
    <>
      <GlobalStyle />
      <Component
        {...pageProps}
        plants={plants}
        toggleFavorite={toggleFavorite}
      />
    </>
  );
}

```

3. `pages/index.js`
- I recognized that you get an error message about the rendering of different CSS classes in this part:

```jsx
        <LocationContainer>
          <LocationSymbol>&#x1F4CD;</LocationSymbol>
          <LocationText>{selectedLocation.name}</LocationText>
          <HamburgerMenu onClick={toggleDropdown}>&#x2630;</HamburgerMenu>
          {isDropdownOpen && (
            <DropdownMenu isOpen={isDropdownOpen}>
              {locations.map((location) => (
                <DropdownItem
                  key={location.id}
                  onClick={() => handleLocationChange(location)}
                >
                  {location.name}
                </DropdownItem>
              ))}
            </DropdownMenu>
          )}
        </LocationContainer>
```
- I do not know yet why, please try to check or we need to investigate further
- I took the props and hand it over to your `PlantsPage`:
```jsx
export default function RegionPlants({ plants, toggleFavorite }) {
...
      <PlantPage plants={plants} toggleFavorite={toggleFavorite} />
    </div>
  );
}
```

4. `PlantPage`
- i did the same for your PlantPage:
```jsx
import PlantList from "./PlantList";

function PlantPage({ plants, toggleFavorite }) {
  return <PlantList plants={plants} toggleFavorite={toggleFavorite} />;
}

export default PlantPage;
```
5. `PlantList`
- removed the SWR import
- I changed your PlantList component to handle the props and added a new prop `id`:

```jsx
function PlantList({ plants, toggleFavorite }) {
  return (
    <List>
      {plants.map((plant) => (
        <li key={plant.id}>
          <PlantCard
            id={plant.id}
            name={plant.name}
            image={plant.image}
            toggleFavorite={toggleFavorite}
          />
        </li>
      ))}
    </List>
  );
}
```

I removed this part:

```jsx
export const fetcher = async (...args) => {
  const response = await fetch(...args);
  if (!response.ok) {
    throw new Error(`Request with ${JSON.stringify(args)} failed.`);
  }
  return await response.json();
};

function PlantList (){
  const {data: plants, error, isLoading} = useSWR("/api/Plants", fetcher);
  if (error){
    return (<p>Error!</p>)
  }
  if (isLoading)
   {return <p>Loading...</p>
}
return (<List>
  {plants.map((plant, index) => (
    <li key={index}>
      <PlantCard
        name={plant.title}
        // image={plant.image}
        onClick={() => navigateToPlantDetails(plant.id)}
      />
    </li>
  ))}
</List>)
}
```

6. `PlantCard`
- I also did a refactoring of your PlantCard component:

```jsx
import { FiHeart as Heart } from "react-icons/fi";

...

function PlantCard({ name, image, toggleFavorite, id }) {
  return (
    <Container>
      <StyledImage
        src={image}
        alt={name}
        fill
        sizes="(max-width: 768px) 100vw,
              (max-width: 1200px) 50vw,
              33vw"
      />
      <Button type="button" onClick={() => toggleFavorite(id)}>
        <Heart />
      </Button>
      <Title>{name}</Title>
    </Container>
  );
}
```
- I moved your Heart Icon the Plant component as I think this makes more sense
- something is not right with you heart icon. therefore I removed the styled component and changed the name of the icon in the import
- also you need to work with a button here, as this is the right interactive element to toggle
- I added a basic styling for the button:

```jsx
const Button = styled.button`
  position: absolute;
`;
```
- I need to investigate further what went wrong with the icon styling but for now this should work out

7. `pages/favorites.js`

- I removed the plants data and instead worked with the plants prop:
```jsx
import FavoritesPage from "@/components/FavoritesPage";

export default function Favorites({ plants }) {
  return <FavoritesPage plants={plants} />;
}
```

8. `FavoritesPage`
- I removed the favorites state as it is not necessary here. A const that saves the filtered favorites is enough
- then your can hand over these favorites to your PlantList and reuse the design you already have for your index page :)

```jsx
function FavoritesPage({ plants }) {
  const router = useRouter();

  const favorites = plants.filter((plant) => plant.isFavorite);

  return (
    <>
      <h1>Favorites</h1>
      <PlantList plants={favorites} />
      <BackButton onClick={() => router.back()}>Back to Gallery</BackButton>
    </>
  );
}
```

I hope this does help. :)
